### PR TITLE
heartbeat + enhanced tests + refactor

### LINKED ISSUE #17: Heartbeat validator message
Periodically sign a Heartbeat event and send it to the other validators; this event might contain some data to distinguish the time, e.g. timestamp or sequence

this would solve #13, help with #6 and help the market determine the health of the channel

more accurately, it allows the market to distinguish between more detailed off-chain state: a `waiting` state where we are waiting for all validators to post a heartbeat, and then `offline` (a type of unhealthy state) which happens if some validator hasn't sent a heartbeat recently; 

once this is done, tests must filter validator messages by type (otherwise the last messages will be full of Heartbeat)

see also: http://github.com/adexnetwork/adex-market/issues/1 and https://github.com/AdExNetwork/adex-protocol/blob/master/components/market.md

### PR DESCRIPTION
**Pull request by samparsky**
Fixes #17

### CODE REVIEW

File: services/validatorWorker/heartbeat.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+function heartbeat(adapter, channel){
+	const whoami = adapter.whoami()
+	const validatorIdx = channel.validators.indexOf(whoami)
+	assert.ok(validatorIdx !== -1, 'validatorTick: sending heartbeat for a channel where we are not validating')
+	const otherValidators = channel.spec.validators.filter(v => v.id != whoami)
+
+	const timestamp = `${new Date().getTime() / 1000}`
+
+	return adapter.sign(timestamp)
```
Comment (Ivshti) [inline services/validatorWorker/heartbeat.js]: I think you should hash the timestamp so we always give the same format to `adapter.sign` 
also this needs to include the channelId too, as it's vulnerable to replay attacks too

perhaps we can use the new function that we added to the adapter as well? getSignableStateRoot

File: test/integration.js [hunk 2]
```diff
@@ -236,7 +242,7 @@ tape('POST /channel/{id}/{validator-messages}: wrong signature', function(t) {
 			"messages": [{ 
 				"type": 'NewState', 
 				stateRoot,
-				"balances": { "myAwesomePublisher" : "9" },
+				"balances": { "myAwesomePublisher" : "12", "anotherPublisher": "3" },
```
Comment (Ivshti) [inline test/integration.js]: why is this changed?

Comment (samparsky) [reply test/integration.js]: The test case was throwing a different error than the required one. It was throwing `invalid transition error` instead of `invalid signature`

File: services/validatorWorker/heartbeat.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+const assert = require('assert')
+const { persistAndPropagate } = require("./lib/propagation")
+
+function heartbeat(adapter, channel){
+	const whoami = adapter.whoami()
+	const validatorIdx = channel.validators.indexOf(whoami)
+	assert.ok(validatorIdx !== -1, 'validatorTick: sending heartbeat for a channel where we are not validating')
+	const otherValidators = channel.spec.validators.filter(v => v.id != whoami)
+
+	const timestamp = `${new Date().getTime() / 1000}`
```
Comment (Ivshti) [inline services/validatorWorker/heartbeat.js]: should just be in ms

File: services/validatorWorker/heartbeat.js (hunk 1 above)
Comment (samparsky) [reply services/validatorWorker/heartbeat.js]: done

File: test/integration.js [hunk 4]
```diff
@@ -226,23 +234,35 @@ tape('POST /channel/{id}/{events,validator-messages}: wrong authentication', fun
...
-				"lastEvAggr": "2019-01-23T09:08:29.959Z",
-				"signature": "Dummy adapter for 6def5a300acb6fcaa0dab3a41e9d6457b5147a641e641380f8cc4bf5308b16fe by awesomeLeader1" 
-			}]
-		}),
+	fetch(`${followerUrl}/channel/${dummyVals.channel.id}/validator-messages/${dummyVals.ids.leader}/NewState?limit=1`)
+	.then(res => res.json())
+	.then(function(res){
+		const { balances } = res.validatorMessages[0].msg
+
+		let incBalances = {}
```
Comment (Ivshti) [inline test/integration.js]: rename to `deceptiveBalances`

File: test/integration.js [hunk 5]
```diff
@@ -226,23 +234,35 @@ tape('POST /channel/{id}/{events,validator-messages}: wrong authentication', fun
...
+		// increase the state tree balance by 1
+		Object.keys(balances).forEach((item) => (incBalances[item] = `${parseInt(balances[item])+1}`))
```
Comment (Ivshti) [inline test/integration.js]: this thing is not needed for this particular test, only for the next one

File: test/integration.js [hunk 6]
```diff
@@ -257,33 +277,45 @@ tape('POST /channel/{id}/{validator-messages}: wrong signature', function(t) {
...
-			}]
-		}),
+	fetch(`${followerUrl}/channel/${dummyVals.channel.id}/validator-messages/${dummyVals.ids.leader}/NewState?limit=1`)
+	.then(res => res.json())
+	.then(function(res) {
+
+		const { balances } = res.validatorMessages[0].msg
+		const fakeBalances = { "publisher": "3" }
+		// increase the state tree balance by 1
+		let incBalances = {}
```
Comment (Ivshti) [inline test/integration.js]: you don't need to increment; the same balances is also a valid state transition, cause it does not violate the OUTPACE rules

File: services/validatorWorker/follower.js [hunk 7]
```diff
@@ -58,7 +68,9 @@ function onNewState(adapter, {channel, balances, newMsg, approveMsg}) {
 			return { nothingNew: true }
 		}
 	})
-	.then(function(){
+	.then(function(res){
+		if(res && res.nothingNew) return res
```
Comment (Ivshti) [inline services/validatorWorker/follower.js]: this seems bad, but it can be mitigated by just deleting the second `.then`

Comment (Ivshti) [reply services/validatorWorker/follower.js]: just remove lines 70, 71

File: services/validatorWorker/follower.js [hunk 8]
```diff
@@ -58,7 +68,9 @@ function onNewState(adapter, {channel, balances, newMsg, approveMsg}) {
 			return { nothingNew: true }
```
Comment (Ivshti) [inline services/validatorWorker/follower.js]: also the res here from `verify()` should be renamed to `isValid`

File: services/validatorWorker/heartbeat.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
+const assert = require('assert')
+const { persistAndPropagate } = require("./lib/propagation")
```
Comment (Ivshti) [inline services/validatorWorker/heartbeat.js]: use single quotes; maybe we can setup eslint at some point

File: services/validatorWorker/heartbeat.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+		return persistAndPropagate(adapter, otherValidators, channel, {
+			type: 'HeartBeat',
+			timestamp,
+			signature,
+			stateRoot,
+		});
+	})
+}
+
+module.exports = { heartbeat }
```
Comment (Ivshti) [inline services/validatorWorker/heartbeat.js]: just `module.exports = heartbeat` as the convention from follower/leader/etc.

File: services/validatorWorker/leader.js [hunk 11]
```diff
@@ -1,14 +1,24 @@
...
 	return producer.tick(channel)
 		.then(
 			res => res.newStateTree ?
 				afterProducer(adapter, res)
 				: { nothingNew: true }
+		).then(
+			res => {
+				if(res && res.nothingNew){
+					// send heartbeat
+					Promise.resolve(
```
Comment (Ivshti) [inline services/validatorWorker/leader.js]: yeah, the heartbeat is returning a promise, so this should just be `return heartbeat(adapter, channel)`

File: test/integration.js [hunk 12]
```diff
@@ -205,6 +220,57 @@ tape('health works correctly', function(t) {
 	.catch(err => t.fail(err))
 })
 
+tape('heartbeat works correctly', function(t){
+	Promise.resolve()
+	.then(() => wait(waitTime)) // wait till a new state is schedule to be produced
+	.then(function() {
```
Comment (Ivshti) [inline test/integration.js]: should be capitalized `Heartbeat` and not `HeartBeat`

File: test/integration.js [hunk 13]
```diff
@@ -205,6 +220,57 @@ tape('health works correctly', function(t) {
...
+		return fetch(`${followerUrl}/channel/${dummyVals.channel.id}/validator-messages/${dummyVals.ids.follower}/HeartBeat?limit=1`)
```
Comment (Ivshti) [inline test/integration.js]: this is very repetitive, can you refactor it; see the tests for submitting invalid events and etc.

File: test/integration.js [hunk 14]
```diff
@@ -226,23 +292,35 @@ tape('POST /channel/{id}/{events,validator-messages}: wrong authentication', fun
 })
 
 tape('POST /channel/{id}/{validator-messages}: wrong signature', function(t) {
-	const stateRoot = "6def5a300acb6fcaa0dab3a41e9d6457b5147a641e641380f8cc4bf5308b16fe"
+	let stateRoot = ""
```
Comment (Ivshti) [inline test/integration.js]: do you have any clue why this was not failing considering we had the bug in `follower`

File: test/integration.js (hunk 13 above)
Comment (samparsky) [reply test/integration.js]: yeah definitely

File: services/validatorWorker/follower.js [hunk 15]
```diff
@@ -28,6 +29,13 @@ function tick(adapter, channel) {
 			return onNewState(adapter, { ...res, newMsg, approveMsg })
 		})
 	})
+	.then(function(res){
+		// send heartbeat
+		if(res && res.nothingNew){
+			heartbeat(adapter, channel)
```
Comment (Ivshti) [inline services/validatorWorker/follower.js]: should be `return heartbeat`, right?
